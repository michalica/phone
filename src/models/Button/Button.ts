
export default class Button {
  constructor(
        private identifier: number,
        private content: string,
    ) {
  }

  public getIdentifier(): number {
    return this.identifier;
  }

  public getContent(): string {
    return this.content;
  }
}
