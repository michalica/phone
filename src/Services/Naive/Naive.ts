import WordDeterminationInterface from "../TNine/WordDetermination.interface";
import Button from "../../models/Button";


/*
*
* Algorithm from here http://mitchrobb.com/blog/Solving-T9-Autocomplete-with-a-Prefix-Tree/
* */

export default class Naive implements WordDeterminationInterface {
    public compute(digits: Button[]): string[]
    {
        return this.getWords(digits);
    }

    private getWords(digits: Button[]): string[]
    {
        let results = [''];

        digits.forEach((button: Button) => {
            results = Naive.addDigit(results, button.getContent())
        })

        return results;
    }

    private static addDigit(array: string[], options: string): string[]
    {
        const newArray = [];
        for(let i = 0; i < array.length; i++) {
            // for all the solutions so far
            const subSolution = array[i];
            for(let x = 0; x < options.length; x++) {
                const letter = options[x];
                // for every letter corresponding to this digit,
                // we create a new array combining that subsolution and the option,
                // and push it into our new result.
                newArray.push(subSolution.concat(letter));
            }
        }
        return newArray;
    }
}
