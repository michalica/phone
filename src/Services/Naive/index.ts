import Naive from "./Naive";

export default Naive;