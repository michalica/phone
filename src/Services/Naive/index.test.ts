import Naive from "./Naive";
import Button from "../../models/Button";

describe('smoke test', () => {
    it('should create instance', () => {
        const instance = new Naive();

        expect(instance).not.toBeNull();
    });
})
describe('snapshot tests', () => {
    let instance: Naive;
    beforeEach(() => {
        instance = new Naive();
    })

    it('compute for one button', () => {

        const button = new Button(2, 'abc');

        expect(instance.compute([button])).toMatchSnapshot();
    });

    it('compute for two buttons', () => {

        const button = new Button(2, 'abc');
        const button2 = new Button(3, 'def');
        expect(instance.compute([button, button2])).toMatchSnapshot();
    });

    it('compute for two buttons', () => {

        const button = new Button(2, 'abc');
        const button2 = new Button(3, 'def');
        const button3 = new Button(3, 'ghi');

        expect(instance.compute([button, button2, button3])).toMatchSnapshot();
    });

})