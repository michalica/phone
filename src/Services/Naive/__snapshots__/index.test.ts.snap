// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests compute for one button 1`] = `
Array [
  "a",
  "b",
  "c",
]
`;

exports[`snapshot tests compute for two buttons 1`] = `
Array [
  "ad",
  "ae",
  "af",
  "bd",
  "be",
  "bf",
  "cd",
  "ce",
  "cf",
]
`;

exports[`snapshot tests compute for two buttons 2`] = `
Array [
  "adg",
  "adh",
  "adi",
  "aeg",
  "aeh",
  "aei",
  "afg",
  "afh",
  "afi",
  "bdg",
  "bdh",
  "bdi",
  "beg",
  "beh",
  "bei",
  "bfg",
  "bfh",
  "bfi",
  "cdg",
  "cdh",
  "cdi",
  "ceg",
  "ceh",
  "cei",
  "cfg",
  "cfh",
  "cfi",
]
`;
